### DISCUSSION TITLE
Support for ES6 modules

### DISCUSSION THREAD

**Discussion by ithilbor** [tailwindlabs/tailwindcss#8282] [Help]
I've followed the guide but when i'm starting the app it gives me this error:

Compiled with problems:

ERROR in ./src/index.css (./node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[5].use[1]!./node_modules/react-scripts/node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[5].use[2]!./node_modules/source-map-loader/dist/cjs.js!./src/index.css)

Module build failed (from ./node_modules/react-scripts/node_modules/postcss-loader/dist/cjs.js):
Error [ERR_REQUIRE_ESM]: require() of ES Module /../client/tailwind.config.js from /../client/node_modules/tailwindcss/lib/lib/setupTrackingContext.js not supported.
tailwind.config.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
Instead rename tailwind.config.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in /../client/package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).

There is a way to use tailwind with ES6 module?
Thanks in advance.

**Comment by skimhugo** [comment]
change the file tailwind.config.js to tailwind.config.cjs

**Reply by ithilbor** [reply]
I've done it but and in this way it works but it will work also for the production? Is it a good way of getting things done?
I also need to rename the file postcss.config.js to cjs?
Tailwind does not provvide support or a solution for es6 modules?
Also in the index.css is still gave me this error "Unknown at rule @tailwindcss(unknownAtRules)" while using:

@tailwind base;
@tailwind components;
@tailwind utilities;

And when i run the app the tailwind css doesn't affect the elements in which i used it.
PS: i'm using CRA with babel and webpack and react v.18.0.0 and nodejs v.v16.14.2
Thanks.

**Reply by jmtornetta** [reply]
I also haven't been able to get tailwindcss stylesheets to generate with my ES module. I tested today by installing a new Next.js app using create-next-app and was able to setup tailwindcss properly with CommonJS (control case).

But, once I changed my node app to an ES module by changing the "type" to "module", my stylesheets wouldn't generate. I tried:
1. Renaming my postcss and tailwind config files to .cjs.
2. Adjusting the config statements to `export default {...}` vs `mode.exports = {...}`
3. Adding `tailwindcss: { config: './tailwind.config.cjs' },` to my `postcss.config.cjs`
4. Creating a symlink to `tailwind.config.cjs` from `tailwind.config.js`

The only way I could get tailwindcss working with my ES module was to use the CLI tool to build my stylesheet output manually. Example: `npx tailwindcss -i in.css -o out.css -c tailwind.config.cjs`. But I can't really design anything like this unless I build a bunch of bash scripts to recreate my CSS any time a change is detected.

Anyone know what's happening and why css isn't being watched or compiled automatically? Or have a workaround (e.g. loading all tailwindcss by default in a development environment)?

**Comment by jmtornetta** [comment]
Okay, I ended up just overriding the default watch behavior by adding the following to my `package.json`. This worked reliably for development and design for me. When I build, I'll  just remove the `--watch` parameter.
```
  "scripts": {
    "dev": "next dev && tailwindcss -i ./styles/input.css -o ./styles/globals.css --watch -c tailwind.config.cjs",
   }
```
I can only guess that the default next.js config doesn't read tailwind.config.cjs for some reason.

**Comment by macarie** [comment]
I had the same problem, and a workaround I've found to be working is using a JSON PostCSS configuration file.

After I've renamed `postcss.config.js` to `postcss.config.json` and `tailwind.config.js` to `tailwind.config.cjs` the styles showed up.

I think Next.js is not picking up the `.cjs` PostCSS configuration file, maybe this is worth mentioning in the docs?

**Comment by LarsArtmann** [comment]
Is there a reason technical reason for not using/exporting ESModules (https://publint.dev/tailwindcss@3.3.3)?

**Comment by doutatsu** [comment]
With Vite 5 / Rollup 5 requiring ES modules, this became a more acute problem now. That includes having to do `const defaultTheme = require('tailwindcss/defaultTheme');` instead of importing it as a module. I'd love to see it being addressed

**Reply by wongjn** [reply]
As of v3.3.0, Tailwind supports ESM syntax for the configuration file.
